pandas/tests/test_frame.py

```diff
@@ -12385,6 +12404,93 @@ def verify(df):
                 for col in ['4th', '5th']:
                     verify(udf[col])
 
+        # GH7403
+        df = pd.DataFrame({'A': list('aaaabbbb'),'B':range(8), 'C':range(8)})
+        df.iloc[3, 1] = np.NaN
+        left = df.set_index(['A', 'B']).unstack(0)
+
+        vals = [[3, 0, 1, 2, nan, nan, nan, nan],
+                [nan, nan, nan, nan, 4, 5, 6, 7]]
+        vals = list(map(list, zip(*vals)))
+        idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')
```

Comment:
> unless you explicitly cast the integers to objects rather than float64

I actually explicitly [cast to objects](https://github.com/pydata/pandas/blob/726e892e22ef85883b135472102ae435388a6196/pandas/core/reshape.py#L238) not float64. the index comes out as float because of type inference in [`Index.__new__`](https://github.com/pydata/pandas/blob/726e892e22ef85883b135472102ae435388a6196/pandas/core/index.py#L195), not in unstacking code.

    >>> i
    array([nan, 1, 2], dtype=object)
    >>> type(i[1])
    <class 'int'>
    >>> Index(i)
    Float64Index([nan, 1.0, 2.0], dtype='float64')